### PR TITLE
Constrain project name during pulumi new

### PR DESCRIPTION
**Pull request by ellismg**
When picking a project name using `pulumi new`, ensure that the name
matches the regexp `^[A-Za-z0-9_.-]{1,100}$`. We'll start enforcing
this at other layers of the stack in the future (for now you could
work around these restrictions by hand editing `Pulumi.yaml`, but for
now, we'll just start restricting this to new projects created with
the wizard.

### CODE REVIEW

File: cmd/new.go [hunk 1]
```diff
@@ -766,17 +767,21 @@ func promptForConfig(
...
 	if yes {
 		return defaultValue, nil
 	}
 
 	for {
+		var prompt string
+
 		if defaultValue == "" {
 			prompt = opts.Color.Colorize(
 				fmt.Sprintf("%s%s:%s ", colors.BrightCyan, prompt, colors.Reset))
```
Comment (justinvp) [inline cmd/new.go]: `prompt` => `valueType` here, right?

```suggestion
				fmt.Sprintf("%s%s:%s ", colors.BrightCyan, valueType, colors.Reset))
```

File: cmd/new.go [hunk 2]
```diff
@@ -200,7 +200,7 @@ func newNewCmd() *cobra.Command {
 			// Prompt for the project name, if it wasn't already specified.
 			if name == "" {
 				defaultValue := workspace.ValueOrSanitizedDefaultProjectName(name, template.ProjectName, filepath.Base(cwd))
-				name, err = promptForValue(yes, "project name", defaultValue, false, workspace.IsValidProjectName, opts.Display)
+				name, err = promptForValue(yes, "project name", defaultValue, false, validateProjectName, opts.Display)
```
[the remark this answers is not in the document]
Comment (ellismg) [reply cmd/new.go]: Great feedback.  I moved these helpers into the workspace package and removed IsValidProjectName, moving all uses over to ValidateProjectName.

[the remark this answers is not in the document]
Comment (ellismg) [reply cmd/new.go]: And great catch!

File: cmd/new.go [hunk 3]
```diff
@@ -766,17 +767,21 @@ func promptForConfig(
 
 // promptForValue prompts the user for a value with a defaultValue preselected. Hitting enter accepts the
 // default. If yes is true, defaultValue is returned without prompting. isValidFn is an optional parameter;
-// when specified, it will be run to validate that value entered. An invalid value will result in an error
-// message followed by another prompt for the value.
+// when specified, it will be run to validate that value entered. When this function returns a non empty string
+// validation is assumed to have failed and an error is printed. The string returned by isValidFn is also displayed
+// to provide information about why the validation failed. A period is appended to this message. `promptForValue` then
+// pompts again.
```
[the remark this answers is not in the document]
Comment (ellismg) [reply cmd/new.go]: Fixed.

File: cmd/new.go (hunk 1 above)
Comment (ellismg) [reply cmd/new.go]: Yup!  Thanks!

**Pull request merged**